File: sample-code/vehicle door.cpp
#include <iostream>

using namespace std;

// Define the states of the door
enum DoorState {
  CLOSED,
  OPENING,
  OPEN,
  CLOSING
};

// Define the door object
class Door {
public:
  DoorState state;
  int position;

  // Constructor
  Door() {
    state = CLOSED;
    position = 0;
    cout<<"Door created"<<endl;
  }

  // Methods to change the state of the door
  void open() {
    state = OPENING;
    position = 100;
    cout<<"Door openened"<<endl;
  }

  void close() {
    state = CLOSING;
    position = 0;
    cout<<"Door closed"<<endl;
  }

  // Method to update the position of the door
  void updatePosition() {
    if (state == OPENING) {
      position += 1;
      cout<<"Updated position: open"<<endl;
    } else if (state == CLOSING) {
      position -= 1;
      cout<<"Updated position: closed"<<endl;
    }
  }
};

// Main function
int main() {
  // Create a door object
  Door door;

  // Open the door
  door.open();

  // Update the position of the door
  door.updatePosition();

  // Close the door
  door.close();

  // Update the position of the door
  door.updatePosition();

  return 0;
}

File: sample-code/vehicle door.py
from enum import Enum

# Define the states of the door


class DoorState(Enum):
    CLOSED = 0
    OPENING = 1
    OPEN = 2
    CLOSING = 3

# Define the door object


class Door:
    def __init__(self):
        self.state = DoorState.CLOSED
        self.position = 0
        print("Door created")

    # Methods to change the state of the door
    def open(self):
        self.state = DoorState.OPENING
        self.position = 100
        print("Door openened")

    def close(self):
        self.state = DoorState.CLOSING
        self.position = 0
        print("Door closed")

    # Method to update the position of the door
    def update_position(self):
        if self.state == DoorState.OPENING:
            self.position += 1
            print("Updated position: open")
        elif self.state == DoorState.CLOSING:
            self.position -= 1
            print("Updated position: closed")

# Main function


def main():
    # Create a door object
    door = Door()

    # Open the door
    door.open()

    # Update the position of the door
    door.update_position()

    # Close the door
    door.close()

    # Update the position of the door
    door.update_position()


if __name__ == "__main__":
    main()
